src/test/java/htsjdk/samtools/LongReadsTest.java

```diff
@@ -0,0 +1,227 @@
+import htsjdk.beta.plugin.reads.ReadsEncoder;
+import htsjdk.beta.plugin.reads.ReadsEncoderOptions;
+import htsjdk.beta.plugin.registry.HtsDefaultRegistry;
+import htsjdk.io.HtsPath;
+import htsjdk.io.IOPath;
+import htsjdk.samtools.util.CloseableIterator;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class LongReadsTest extends HtsjdkTest {
+    final String TEST_DATA = "src/test/resources/htsjdk/samtools/longreads/";
+
+    // samtools-generated BAM of PacBio-sequenced reads, containing a subset of chr1, plus a handful of unmapped reads
+    final IOPath PAC_BIO_READS = new HtsPath(TEST_DATA +
+            "NA12878.m64020_190210_035026.chr21.5011316.5411316.unmapped.bam");
+
+    @Test
+    final public void testRoundTripBAM() throws IOException {
+        // create a copy of the test bam, with an index, and then compare the header and all reads from
+        // the copy with the original
+        final IOPath copiedBAMPath = createBAMCopyWithIndex(PAC_BIO_READS);
+
+        final ReadsDecoderOptions readsDecoderOptions = new ReadsDecoderOptions()
+                .setValidationStringency(ValidationStringency.STRICT)
+                .setDecodeEagerly(true);
+
+        try (final ReadsDecoder originalDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(PAC_BIO_READS, readsDecoderOptions);
+             final ReadsDecoder copyDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(copiedBAMPath, readsDecoderOptions);
+             final CloseableIterator<SAMRecord> originalIt = originalDecoder.iterator();
+             final CloseableIterator<SAMRecord> copyIt = copyDecoder.iterator()) {
+
+            Assert.assertEquals(copyDecoder.getHeader(), originalDecoder.getHeader());
+
+            while (originalIt.hasNext()) {
+                final SAMRecord originalRecord = originalIt.next();
+                Assert.assertTrue(copyIt.hasNext());
+                final SAMRecord copyRecord = copyIt.next();
+
+                // validate
+                Assert.assertNull(originalRecord.isValid());
+                Assert.assertNull(copyRecord.isValid());
+                Assert.assertEquals(copyRecord, originalRecord);
+            }
+            Assert.assertFalse(copyIt.hasNext());
+        }
+    }
+
+    @Test
+    final public void testQueryOverlappingForEachRead() throws IOException {
+        final IOPath copiedBAMPath = createBAMCopyWithIndex(PAC_BIO_READS);
+
+        // for each mapped read in the original source BAM, query both the original and the copy for
+        // all reads that overlap that read's coordinates, and compare the results
+        try (final ReadsDecoder sourceDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(ReadsBundle.resolveIndex(PAC_BIO_READS));
```

Comment:
This is an aside, but ReadsBundle.resolveIndex() probably needs a bit of renaming.  Seeing it in practice makes me think it should be ReadsBundle.withResolvedIndex() or something along those lines to make it clear that it's retuning a new reads bundle.

Reply:
Yes, makes sense.